QA - using spark connector to unload tables uses all memory and never finishes for big tables

Already confirmed by Andrei. Jdbc unload should not try to take all memory. 
It always seem to fail with:

testcase:
took 7gb lineitem table, loaded it. Tried to unload into parquet file. Always hangs no matter which memory settings I give to the Shell. 

spark-shell --master yarn --conf "spark.executor.memory=8G" --conf "spark.driver.memory=8G" --jars /home/actian/kelch01/Spark/spark-loader/spark_vector_loader-assembly-1.0-SNAPSHOT.jar
 
sqlContext.sql("""CREATE TEMPORARY TABLE lineitem
USING  com.actian.spark_vector.sql.DefaultSource
OPTIONS (
    host "uksl-kelch01-cent6-clu1",
    instance "K2",
    database "testdb",
    table "lineitem"
)""")

sqlContext.sql("select * from lineitem").write.parquet("hdfs://uksl-kelch01-cent6-clu1.actian.com:8020/Actian/tmp/lineitem.parquet")


[ERROR] [02/25/2016 15:15:48.155] [sparkDriver-scheduler-1] [ActorSystem(sparkDriver)] exception on LARS’ timer thread
java.lang.OutOfMemoryError: GC overhead limit exceeded

[ERROR] [02/25/2016 15:16:26.251] [sparkDriver-scheduler-1] [ActorSystem(sparkDriver)] Uncaught fatal error from thread [sparkDriver-scheduler-1] shutting down ActorSystem [sparkDriver]
java.lang.OutOfMemoryError: GC overhead limit exceeded

java.lang.OutOfMemoryError: GC overhead limit exceeded

--- a/src/main/scala/com/actian/spark_vector/reader/DataStreamReader.scala
+++ b/src/main/scala/com/actian/spark_vector/reader/DataStreamReader.scala
@@ -13,7 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.actian.spark_vector.writer
+package com.actian.spark_vector.reader
 
 import scala.annotation.tailrec
 import java.nio.ByteBuffer

--- a/src/main/scala/com/actian/spark_vector/reader/ScanRDD.scala
+++ b/src/main/scala/com/actian/spark_vector/reader/ScanRDD.scala
@@ -52,7 +52,7 @@ class ScanRDD(
       close(stmt, "statement")
       close(cxn, "vector JDBC connection")
       closed = true
-      if (it != null) it.profilePrint(it.prof_accs)
+      if (it != null) it.profilePrint(it.profAccs)
     }
 
     logDebug(s"Computing partition in ScanRDD by issuing JDBC select statement: $vectorSelectStatement")

--- a/src/main/scala/com/actian/spark_vector/reader/ScanRDD.scala
+++ b/src/main/scala/com/actian/spark_vector/reader/ScanRDD.scala
@@ -22,10 +22,10 @@ import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.Row
 import org.apache.spark.sql.execution.datasources.jdbc.DriverRegistry
 
-import com.actian.spark_vector.vector.{ ResultSetToRowIterator, VectorConnectionProperties, VectorJDBC }
+import com.actian.spark_vector.vector.{ ResultSetRowIterator, VectorConnectionProperties, VectorJDBC }
 
 /**
- * `Vector` RDD to load data into `Spark` through its `DataStream API`
+ * `Vector` RDD to load data into `Spark` through JDBC
  *
  *  @param connProps contains the read configuration needed to connect to the Vector front-end
  *  @param vectorSelectStatement Vector prepared SQL statement to be issued through JDBC to start exporting data
@@ -37,7 +37,8 @@ class ScanRDD(
     vectorSelectStatement: String,
     vectorSelectParams: Seq[Any] = Nil) extends RDD[Row](sc, Nil) with Logging {
   private var closed = false
-  private var it: ResultSetToRowIterator = null
+  private var it: ResultSetRowIterator = null
+
   override protected def getPartitions = Array(new Partition { def index = 0 })
 
   override protected def getPreferredLocations(split: Partition) = Seq(connProps.host)
@@ -57,11 +58,11 @@ class ScanRDD(
     logDebug(s"Computing partition in ScanRDD by issuing JDBC select statement: $vectorSelectStatement")
     lazy val cxn = new VectorJDBC(connProps)
     lazy val (stmt, rs) = cxn.executeUnmanagedPreparedQuery(vectorSelectStatement, vectorSelectParams)
-    it = new ResultSetToRowIterator(rs)
+    it = new ResultSetRowIterator(rs)
     it
   }
 
-  private def close[T <: { def close() }](c: T, resourceName: String): Unit = if (!closed) {
-    try if (c != null) { c.close } catch { case e: Exception => logWarning(s"Exception closing $resourceName", e) }
+  private def close[T <: { def close() }](c: T, resourceName: String): Unit = if (!closed && c != null) {
+    try { c.close } catch { case e: Exception => logWarning(s"Exception closing $resourceName", e) }
   }
 }

--- a/src/main/scala/com/actian/spark_vector/reader/ScanRDD.scala
+++ b/src/main/scala/com/actian/spark_vector/reader/ScanRDD.scala
@@ -1,0 +1,67 @@
+/*
+ * Copyright 2016 Actian Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.actian.spark_vector.reader
+
+import scala.language.reflectiveCalls
+
+import org.apache.spark.{ Logging, Partition, SparkContext, TaskContext }
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.execution.datasources.jdbc.DriverRegistry
+
+import com.actian.spark_vector.vector.{ ResultSetToRowIterator, VectorConnectionProperties, VectorJDBC }
+
+/**
+ * `Vector` RDD to load data into `Spark` through its `DataStream API`
+ *
+ *  @param connProps contains the read configuration needed to connect to the Vector front-end
+ *  @param vectorSelectStatement Vector prepared SQL statement to be issued through JDBC to start exporting data
+ *  @param vectorSelectParams  Parameters to the prepared SQL statement, if any
+ */
+class ScanRDD(
+    @transient private val sc: SparkContext,
+    connProps: VectorConnectionProperties,
+    vectorSelectStatement: String,
+    vectorSelectParams: Seq[Any] = Nil) extends RDD[Row](sc, Nil) with Logging {
+  private var closed = false
+  private var it: ResultSetToRowIterator = null
+  override protected def getPartitions = Array(new Partition { def index = 0 })
+
+  override protected def getPreferredLocations(split: Partition) = Seq(connProps.host)
+
+  override def compute(split: Partition, task: TaskContext): Iterator[Row] = {
+    /* Need to register ingres JDBC driver so that it is picked up on the executors */
+    DriverRegistry.register(VectorJDBC.DriverClassName)
+
+    task.addTaskCompletionListener { _ =>
+      close(rs, "result set")
+      close(stmt, "statement")
+      close(cxn, "vector JDBC connection")
+      closed = true
+      if (it != null) it.profilePrint(it.prof_accs)
+    }
+
+    logDebug(s"Computing partition in ScanRDD by issuing JDBC select statement: $vectorSelectStatement")
+    lazy val cxn = new VectorJDBC(connProps)
+    lazy val (stmt, rs) = cxn.executeUnmanagedPreparedQuery(vectorSelectStatement, vectorSelectParams)
+    it = new ResultSetToRowIterator(rs)
+    it
+  }
+
+  private def close[T <: { def close() }](c: T, resourceName: String): Unit = if (!closed) {
+    try if (c != null) { c.close } catch { case e: Exception => logWarning(s"Exception closing $resourceName", e) }
+  }
+}

--- a/src/main/scala/com/actian/spark_vector/sql/VectorRelation.scala
+++ b/src/main/scala/com/actian/spark_vector/sql/VectorRelation.scala
@@ -21,11 +21,12 @@ import org.apache.spark.sql.{ DataFrame, Row, SQLContext, sources }
 import org.apache.spark.sql.sources.{ BaseRelation, Filter, InsertableRelation, PrunedFilteredScan }
 import org.apache.spark.sql.types.StructType
 
-import com.actian.spark_vector.vector.VectorJDBC;
+import com.actian.spark_vector.reader.ScanRDD
+import com.actian.spark_vector.vector.{ VectorJDBC, VectorOps }
 
 private[sql] class VectorRelation(tableRef: TableRef, userSpecifiedSchema: Option[StructType], override val sqlContext: SQLContext)
     extends BaseRelation with InsertableRelation with PrunedFilteredScan with Logging {
-  import com.actian.spark_vector.vector.VectorOps._
+  import VectorOps._
 
   override def schema: StructType =
     userSpecifiedSchema.getOrElse(VectorRelation.structType(tableRef))
@@ -58,18 +59,11 @@ private[sql] class VectorRelation(tableRef: TableRef, userSpecifiedSchema: Optio
     val (whereClause, whereParams) = VectorRelation.generateWhereClause(filters)
     val statement = s"select $columns from ${tableRef.table} $whereClause"
     logDebug(s"Executing Vector select statement: $statement")
-    /** TODO: replace this JDBC call with creating a RDD that reads data in parallel from `Vector` in parallel */
-    val results = VectorJDBC.withJDBC(tableRef.toConnectionProps) { cxn =>
-      cxn.query(statement, whereParams);
-    }
-
-    val rows = results.map(Row.fromSeq(_))
-    sqlContext.sparkContext.parallelize(rows)
+    new ScanRDD(sqlContext.sparkContext, tableRef.toConnectionProps, statement, whereParams)
   }
 }
 
 object VectorRelation {
-
   def apply(tableRef: TableRef, userSpecifiedSchema: Option[StructType], sqlContext: SQLContext): VectorRelation = {
     new VectorRelation(tableRef, userSpecifiedSchema, sqlContext)
   }
@@ -115,9 +109,10 @@ object VectorRelation {
    */
   def generateWhereClause(filters: Array[Filter]): (String, Seq[Any]) = {
     val convertedFilters = filters.map(convertFilter)
-    if (!convertedFilters.isEmpty)
+    if (!convertedFilters.isEmpty) {
       (convertedFilters.map(_._1).mkString("where ", " and ", ""), convertedFilters.flatMap(_._2))
-    else
+    } else {
       ("", Nil)
+    }
   }
 }

--- a/src/main/scala/com/actian/spark_vector/vector/LoadVector.scala
+++ b/src/main/scala/com/actian/spark_vector/vector/LoadVector.scala
@@ -24,7 +24,7 @@ import org.apache.spark.sql.types.StructType
 
 import com.actian.spark_vector.util.{ RDDUtil, ResourceUtil }
 import com.actian.spark_vector.vector.Vector._
-import com.actian.spark_vector.writer.{ DataStreamRDD, DataStreamWriter, RowWriter }
+import com.actian.spark_vector.writer.{ DataStreamWriter, InsertRDD, RowWriter }
 
 /** Utility object that defines methods for loading data into Vector */
 object LoadVector extends Logging {
@@ -76,9 +76,9 @@ object LoadVector extends Logging {
       preSQL.foreach(_.foreach(writer.client.getJdbc.executeStatement))
     }
 
-    val datastreamRDD = new DataStreamRDD(finalRDD, writer.writeConf)
+    val insertRDD = new InsertRDD(finalRDD, writer.writeConf)
     val result = writer.initiateLoad
-    datastreamRDD.sparkContext.runJob(datastreamRDD, writer.write _)
+    insertRDD.sparkContext.runJob(insertRDD, writer.write _)
     // FIX ME
     val rowCount = Await.result(result, Duration.Inf)
     if (rowCount >= 0) {

--- a/src/main/scala/com/actian/spark_vector/vector/VectorJDBC.scala
+++ b/src/main/scala/com/actian/spark_vector/vector/VectorJDBC.scala
@@ -15,12 +15,49 @@
  */
 package com.actian.spark_vector.vector
 
-import java.sql._
+import java.sql.{ Connection, DriverManager, PreparedStatement, ResultSet, Statement }
 import java.util.Properties
-import com.actian.spark_vector.util.ResourceUtil
-import com.actian.spark_vector.vector.ErrorCodes._
+
 import org.apache.spark.Logging
-import com.actian.spark_vector.sql.VectorRelation
+import org.apache.spark.sql.Row
+
+import com.actian.spark_vector.Profiling
+import com.actian.spark_vector.util.ResourceUtil.RichExtractableManagedResource
+import com.actian.spark_vector.vector.ErrorCodes.{ InvalidDataType, NoSuchTable, SqlException, SqlExecutionError }
+
+import resource.managed
+
+/** Iterator over an ResultSet */
+abstract class ResultSetIterator[T](result: ResultSet) extends Iterator[T] {
+  def extractor: ResultSet => T
+  override def hasNext: Boolean = result.next()
+  override def next(): T = extractor(result)
+}
+
+object ResultSetIterator {
+  def apply[T](result: ResultSet)(ex: ResultSet => T): ResultSetIterator[T] = new ResultSetIterator[T](result) {
+    def extractor = ex
+  }
+}
+
+/** Extracts `Rows` out of a `ResultSet` */
+class ResultSetToRowIterator(result: ResultSet) extends ResultSetIterator[Row](result) with Logging with Profiling {
+  lazy val numColumns = result.getMetaData.getColumnCount
+  lazy val row = collection.mutable.IndexedSeq.fill[Any](numColumns)(null)
+  implicit lazy val prof_accs = profileInit("resultSet extraction")
+  var col = 0
+  def extractor: ResultSet => Row = { rs =>
+    profile("resultSet extraction")
+    col = 0
+    while (col < numColumns) {
+      row(col) = result.getObject(col + 1)
+      col += 1
+    }
+    val ret = Row.fromSeq(row)
+    profileEnd
+    ret
+  }
+}
 
 /**
  * Encapsulate functions for accessing Vector using JDBC
@@ -28,8 +65,8 @@ import com.actian.spark_vector.sql.VectorRelation
 class VectorJDBC(cxnProps: VectorConnectionProperties) extends Logging {
 
   import resource._
-  import ResourceUtil._
-  import VectorRelation._
+  import VectorJDBC._
+  import com.actian.spark_vector.sql.VectorRelation.quote
 
   private implicit class SparkPreparedStatement(statement: PreparedStatement) {
     def setParams(params: Seq[Any]): PreparedStatement = {
@@ -77,6 +114,13 @@ class VectorJDBC(cxnProps: VectorConnectionProperties) extends Logging {
    */
   def executeQuery[T](sql: String)(op: ResultSet => T): T =
     withStatement(statement => managed(statement.executeQuery(sql)).map(op)).resolve()
+
+  /** Execute a SQL query, transferring closing responsibility of the `Statement` and `ResultSet` to the caller */
+  def executeUnmanagedPreparedQuery(sql: String, params: Seq[Any]): (Statement, ResultSet) = {
+    val stmt = dbCxn.prepareStatement(sql)
+    val rs = stmt.setParams(params).executeQuery
+    (stmt, rs)
+  }
 
   /**
    * Execute a prepared `SQL` query closing resources on failures, using scala-arm's `resource` package,
@@ -147,17 +191,10 @@ class VectorJDBC(cxnProps: VectorConnectionProperties) extends Logging {
     })
   }
 
-  private class ResultSetIterator[T](result: ResultSet)(extractor: ResultSet => T) extends Iterator[T] {
-    override def hasNext: Boolean = result.next()
-    override def next(): T = extractor(result)
-  }
-
-  private def toRow(result: ResultSet): Seq[Any] = (1 to result.getMetaData.getColumnCount).map(result.getObject)
-
   /** Execute a select query on Vector and return the results as a matrix of elements */
-  def query(sql: String): Seq[Seq[Any]] = executeQuery(sql)(resultSet => new ResultSetIterator(resultSet)(toRow).toVector)
+  def query(sql: String): Seq[Seq[Any]] = executeQuery(sql)(ResultSetIterator(_)(toRow).toVector)
 
-  def query(sql: String, params: Seq[Any]): Seq[Seq[Any]] = executePreparedQuery(sql, params)(resultSet => new ResultSetIterator(resultSet)(toRow).toVector)
+  def query(sql: String, params: Seq[Any]): Seq[Seq[Any]] = executePreparedQuery(sql, params)(ResultSetIterator(_)(toRow).toVector)
 
   /** Drop the Vector table `tableName` if it exists */
   def dropTable(tableName: String): Unit = {
@@ -193,9 +230,11 @@ class VectorJDBC(cxnProps: VectorConnectionProperties) extends Logging {
 }
 
 object VectorJDBC extends Logging {
-
   import resource._
-  import ResourceUtil._
+
+  final val DriverClassName = "com.ingres.jdbc.IngresDriver"
+
+  def toRow(result: ResultSet): Seq[Any] = (1 to result.getMetaData.getColumnCount).map(result.getObject)
 
   /** Create a `VectorJDBC`, execute the statements specified by `op` and close the `JDBC` connections when finished */
   def withJDBC[T](cxnProps: VectorConnectionProperties)(op: VectorJDBC => T): T = {

--- a/src/main/scala/com/actian/spark_vector/vector/VectorJDBC.scala
+++ b/src/main/scala/com/actian/spark_vector/vector/VectorJDBC.scala
@@ -29,26 +29,25 @@ import resource.managed
 
 /** Iterator over an ResultSet */
 abstract class ResultSetIterator[T](result: ResultSet) extends Iterator[T] {
-  def extractor: ResultSet => T
+  protected def extractor: ResultSet => T
   override def hasNext: Boolean = result.next()
   override def next(): T = extractor(result)
 }
 
 object ResultSetIterator {
   def apply[T](result: ResultSet)(ex: ResultSet => T): ResultSetIterator[T] = new ResultSetIterator[T](result) {
-    def extractor = ex
+    override protected def extractor = ex
   }
 }
 
 /** Extracts `Rows` out of a `ResultSet` */
-class ResultSetToRowIterator(result: ResultSet) extends ResultSetIterator[Row](result) with Logging with Profiling {
+class ResultSetRowIterator(result: ResultSet) extends ResultSetIterator[Row](result) with Logging with Profiling {
   lazy val numColumns = result.getMetaData.getColumnCount
   lazy val row = collection.mutable.IndexedSeq.fill[Any](numColumns)(null)
-  implicit lazy val prof_accs = profileInit("resultSet extraction")
-  var col = 0
-  def extractor: ResultSet => Row = { rs =>
+  implicit lazy val profAccs = profileInit("resultSet extraction")
+  override protected def extractor: ResultSet => Row = { _ =>
     profile("resultSet extraction")
-    col = 0
+    var col = 0
     while (col < numColumns) {
       row(col) = result.getObject(col + 1)
       col += 1
@@ -115,7 +114,7 @@ class VectorJDBC(cxnProps: VectorConnectionProperties) extends Logging {
   def executeQuery[T](sql: String)(op: ResultSet => T): T =
     withStatement(statement => managed(statement.executeQuery(sql)).map(op)).resolve()
 
-  /** Execute a SQL query, transferring closing responsibility of the `Statement` and `ResultSet` to the caller */
+  /** Execute a prepared `SQL` query, transferring closing responsibility of the `Statement` and `ResultSet` to the caller */
   def executeUnmanagedPreparedQuery(sql: String, params: Seq[Any]): (Statement, ResultSet) = {
     val stmt = dbCxn.prepareStatement(sql)
     val rs = stmt.setParams(params).executeQuery

--- a/src/main/scala/com/actian/spark_vector/writer/DataStreamConnector.scala
+++ b/src/main/scala/com/actian/spark_vector/writer/DataStreamConnector.scala
@@ -15,15 +15,12 @@
  */
 package com.actian.spark_vector.writer
 
-import java.io.{ ByteArrayOutputStream, DataOutputStream }
 import java.net.InetSocketAddress
-import java.nio.ByteBuffer
 import java.nio.channels.SocketChannel
-
-import scala.annotation.tailrec
 
 import org.apache.spark.Logging
 
+import com.actian.spark_vector.reader.DataStreamReader
 import com.actian.spark_vector.util.ResourceUtil.closeResourceAfterUse
 import com.actian.spark_vector.writer.srp.VectorSRPClient
 

--- a/src/main/scala/com/actian/spark_vector/writer/DataStreamWriter.scala
+++ b/src/main/scala/com/actian/spark_vector/writer/DataStreamWriter.scala
@@ -94,7 +94,7 @@ class DataStreamWriter[T <% Seq[Any]](
   }
 
   /**
-   * This function is executed once for each partition of `DataStreamRDD` and it will open a socket connection, process all data
+   * This function is executed once for each partition of [[InsertRDD]] and it will open a socket connection, process all data
    * assigned to its corresponding partition (`taskContext.partitionId`) and then close the connection.
    */
   def write(taskContext: TaskContext, data: Iterator[T]): Unit = {

--- a/src/main/scala/com/actian/spark_vector/writer/InsertRDD.scala
+++ b/src/main/scala/com/actian/spark_vector/writer/InsertRDD.scala
@@ -29,7 +29,7 @@ import com.actian.spark_vector.vector.DataStreamPartition
  *  @param rdd `RDD` to be loaded
  *  @param writeConf contains the write configuration needed to connect to `Vector DataStream`s
  */
-class DataStreamRDD[R: ClassTag](
+class InsertRDD[R: ClassTag](
   @transient val rdd: RDD[R],
   writeConf: WriteConf) extends RDD[R](rdd.context, Nil) with Logging {
 

--- a/src/main/scala/com/actian/spark_vector/writer/package.scala
+++ b/src/main/scala/com/actian/spark_vector/writer/package.scala
@@ -24,10 +24,10 @@ package com.actian.spark_vector
  *  `create table` SQL statement that will be first submitted to `Vector`.
  *   - helpers [[RowWriter]] and [[DataStreamWriter]] objects are created and they contain all the needed information for a `Spark` worker to be able to process,
  *  serialize and communicate binary data to `Vector` end points.
- *   - a [[DataStreamRDD]] is created, containing as many partitions as there are `DataStreams` and that will create a `NarrowDependency` to the input `RDD`
+ *   - an [[InsertRDD]] is created, containing as many partitions as there are `DataStreams` and that will create a `NarrowDependency` to the input `RDD`
  *   - driver initiates the load, issuing a SQL query to `Vector` leader node
  *   - driver initiates Spark job => [[RowWriter]] and [[DataStreamWriter]] objects, part of the closure, are serialized and sent to worker processes
- *   - each worker process reads its corresponding write configuration and starts processing input data (as assigned by the driver when [[DataStreamRDD]] was created), serializes it into
+ *   - each worker process reads its corresponding write configuration and starts processing input data (as assigned by the driver when [[InsertRDD]] was created), serializes it into
  *  `ByteBuffers` and then flushes them through the socket towards one (and only one) predetermined `Vector` end point
  *   - during this time, the driver remains blocked waiting for the SQL query to finish. Once all workers are done, the driver then issues a `commit` or `abort` depending on whether any of the
  *  workers failed. Note, we currently do not retry `Spark` workers since partial loading is not supported in `Vector` yet.

--- a/src/main/scala/com/actian/spark_vector/writer/srp/VectorSRPClient.scala
+++ b/src/main/scala/com/actian/spark_vector/writer/srp/VectorSRPClient.scala
@@ -19,7 +19,8 @@ import java.nio.channels.SocketChannel
 import scala.BigInt
 import scala.annotation.tailrec
 import org.apache.spark.Logging
-import com.actian.spark_vector.writer.{ DataStreamReader, DataStreamWriter }
+import com.actian.spark_vector.writer.DataStreamWriter
+import com.actian.spark_vector.reader.DataStreamReader
 import com.actian.spark_vector.vector.ErrorCodes._
 import com.actian.spark_vector.vector.VectorException
 

